### PR TITLE
Ensure IN() Conditions Never Mix Data Types

### PR DESCRIPTION
**Pull request by miqrogroove**
Fixes a MySQL performance issue related to #5266 and simplifies work started by annando to that end.

### CODE REVIEW

File: include/dba.php [hunk 1]
```diff
@@ -1299,37 +1299,29 @@ public static function buildCondition(array &$condition = [])
...
+						 * Workaround for MySQL Bug #64791.
+						 * Never mix data types inside any IN() condition.
+						 * In case of mixed types, cast all as string.
+						 * Logic needs to be consistent with dba::p() data types.
+						 */
 						$is_int = false;
-						$is_float = false;
 						$is_alpha = false;
 						foreach ($value as $single_value) {
 							if (is_int($single_value)) {
```
Comment (MrPetovan) [inline include/dba.php]: This isn't just about integer, float value are concerned as well. If you do rewrite what @annando did earlier, please make it type agnostic. My first hunch would be to take the type of the first element of the array and check the type of the subsequent elements against it. At the first difference, the loop quits and casts everything as a string.

**Comment by miqrogroove** [comment]
No changes needed.  This patch restores the old behavior when mixed types are detected.  There is no casting of floats needed.

**Comment by annando** [comment]
The greatest difference between your approach and my is that I'm casting everything as integer whenever it is possible to do so, while you - in doubt - cast it as string. 

I would like to see performance  tests on this. When casting as string isn't slower, we simply could convert parameters in "IN(...)" at any time - without the need to even check for the cast of every parameter.

File: include/dba.php [hunk 2]
```diff
@@ -1299,37 +1299,29 @@ public static function buildCondition(array &$condition = [])
 						$condition_string .= " AND ";
 					}
 					if (is_array($value)) {
-						// Check if there are integer values in the parameters
+						/**
```
Comment (rabuzarus) [inline include/dba.php]: Code standards: please don't use `/**` for comments because it is used for doxygen. Rather use `/*` or `/* *`

**Comment by MrPetovan** [comment]
> No changes needed. This patch restores the old behavior when mixed types are detected. There is no casting of floats needed.

Why not? A mixed type array like `[10.1, '10.2']` wouldn't trigger the string casting and would likely lead MySQL to ignore the index as well.

**Comment by miqrogroove** [comment]
Negative.  The literal `10.1` returns false for is_int() with dba::p() and both params get bound as string.

**Comment by MrPetovan** [comment]
According to https://github.com/friendica/friendica/pull/5272/files#diff-6cb3a981df757a28e6d6b23c3837dc85R1317

The string casting only happens if at least an integer and a non-integer is found in the value array. A float wouldn't trigger the `is_int` check.

**Comment by miqrogroove** [comment]
Correct.  It has worked that way forever, and this patch restores the behavior.  All non-ints are bound as strings.  And now that we bind ints as ints, we have to make sure not to mix ints with non-ints.  Floats were never at issue here.  All we wanted was to be able to `LIMIT ?`

**Comment by miqrogroove** [comment]
https://github.com/friendica/friendica/blob/423461f844548347da9c7f25f3b92dcb9d00bfc9/include/dba.php#L433

**Comment by MrPetovan** [comment]
Oh, I see what you mean. Nevertheless, this is a potential liability if we want to support more native type binding (not just integers) in the future. I'd rather have a universal mixed type handler that doesn't depend on any ulterior code.

**Comment by miqrogroove** [comment]
I don't think I can help you with a "universal mixed type handler."  I would encourage you to either accept this patch or revert the previous patches, and keep this issue as simple as possible.

**Comment by MrPetovan** [comment]
Follow-up to #5269

**Pull request merged**